torch/csrc/jit/script/compiler.cpp

```diff
@@ -830,23 +939,67 @@ struct to_ir {
   }
 
   Value* emitTernaryIf(const TernaryIf& expr) {
+    auto bool_info = findRefinements(expr.cond());
     Value* cond_value = emitCond(expr.cond());
-    auto true_expr = [&] { return emitExpr(expr.true_expr()); };
-    auto false_expr = [&] { return emitExpr(expr.false_expr()); };
+    auto true_expr = [&] {
+      insertRefinements(bool_info.true_refinements);
+      return emitExpr(expr.true_expr());
+    };
+    auto false_expr = [&] {
+      insertRefinements(bool_info.false_refinements);
+      return emitExpr(expr.false_expr());
+    };
     return emitIfExpr(expr.range(), cond_value, true_expr, false_expr);
   }
 
+  void insertRefinements(const Refinements& ref) {
+    const auto& types = ref.types;
+    const auto& ranges = ref.ranges;
+
+    for (auto value_type: types) {
+      const std::string& name = value_type.first;
+      auto type = value_type.second;
+      auto range = *ranges.at(name);
+      Value * v = environment_stack->getVar(name, range);
+      if (type != NoneType::get()) {
```

Comment:
If it is none, shouldn't it also remap `name` to `None`?

Reply:
Because None currently does not subtype Optionals, this breaks type unification when the if node tries to emit an output.

Reply:
It causes jitter for a reason i haven't looked into. I am going to leave it out on this PR and revisit when None subtypes optional

```
 -     opt_5, x_3 = annotate(Optional[Tensor], None), x_1
?                           ---------      -
+     opt_5, x_3 = annotate(Tensor, None), x_1
```